File: src/lang/java/Comb.java

class Comb {

    long[] fac;
    long[] inv;
    int MOD = (int) 1e9 + 7;
    int size;

    public Comb(int size) {
        this.size = size;
        fac = new long[size + 1];
        inv = new long[size + 1];
        fac[0] = inv[0] = 1;
        for (int i = 1; i <= size; i++) {
            fac[i] = fac[i - 1] * i % MOD;
            // 快速幂计算乘法逆元
            inv[i] = fastPow(fac[i], MOD - 2);
        }
    }

    public long comb(int n, int k) {
        if (n > size || k > size) {
            throw new Error("超出范围");
        }
        return fac[n] * inv[k] % MOD * inv[n - k] % MOD;
    }

    private long fastPow(long a, long b) {
        long base = a % MOD, res = 1;
        while (b != 0) {
            if ((b & 1) == 1) {
                res = res * base % MOD;
            }
            base = base * base % MOD;
            b >>>= 1;
        }
        return res % MOD;
    }
}

File: src/lang/python/comb.py
class Comb:
    def __init__(self, size: int):
        self.MOD = 10**9 + 7
        self.fac = [1] * (size + 1)
        self.inv = [1] * (size + 1)
        self._init(size)

    def _fast_pow(self, a: int, n: int) -> int:
        res = 1
        base = a % self.MOD
        while n > 0:
            if n & 1:
                res = (res * base) % self.MOD
            base = (base * base) % self.MOD
            n >>= 1
        return res

    def _init(self, size: int):
        for i in range(1, size + 1):
            self.fac[i] = self.fac[i - 1] * i % self.MOD
            self.inv[i] = self._fast_pow(self.fac[i], self.MOD - 2)

    def comb(self, n: int, k: int) -> int:
        if k < 0 or k > n:
            return 0
        return (self.fac[n] * self.inv[k] % self.MOD) * self.inv[n - k] % self.MOD
